### DISCUSSION TITLE
Good hosting solution for an FTP server with ftp-srv

### DISCUSSION THREAD

**Discussion by leolabs** [QuorumDMS/ftp-srv#279] [Q&A]
This one is more of a question than an issue.

We've created a custom FTP server that talks to Google Cloud Storage. Currently, we're running it inside of a Docker container on a Google Compute instance, which works alright, but it's a hassle to implement proper Continuous Deployment.

I've been looking for alternatives that offer a better CD experience but it seems like most of them don't offer opening up the ports required to run an FTP server. I've been looking at Heroku, Fly, Render, and Sloppy so far, but neither of them seem to work for what I'm trying to do.

Do you have any recommendations on where to host a project like this? 

Thanks in advance!

**Comment by matt-forster** [comment]
Yea, this is actually a harder problem than it seems on the surface given FTP and its port requirements, as you have discovered.

We currently use [Packer ](https://www.packer.io/) to create an (AWS) AMI and manage our FTP server outside of our containerized environment - this gives us far more control over the networking conditions and dynamic port assignment, etc.

We build the AMI alongside our other container builds and then use [Terraform](https://www.terraform.io/) with an input variable for the AMI build (written to the input file after the tests pass and the container is built) to provision and version the server. This allows us to correctly manage the code within our mono-repo alongside the containerized services and lambda functions and follow the same QA as the rest of the containers.

I'm going to transfer this to the discussions because that is the appropriate place for this!

**Reply by leolabs** [reply]
Thanks, Matt! I'll take a look into Packer and Terraform to see how they can help :)
